File: tests/vectors-cpp/vectors_test.cpp
#include <algorithm>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <functional>
#include <new>
#include <iostream>
#include <map>
#include <sstream>
#include <string>
#include <vector>

#include "generated/device_state_edge.hpp"
#include "generated/device_state_unity.hpp"
#include "generated/every_primitive_edge.hpp"
#include "generated/net_schema.hpp"
#include "generated/player_edge.hpp"
#include "generated/player_info_edge.hpp"
#include "generated/team_edge.hpp"

using namespace generated;

namespace {

std::size_t allocations = 0;

}

void* operator new(std::size_t size) {
    ++allocations;
    if (void* memory = std::malloc(size == 0 ? 1 : size)) return memory;
    throw std::bad_alloc();
}

void operator delete(void* memory) noexcept { std::free(memory); }

void operator delete(void* memory, std::size_t) noexcept { std::free(memory); }

namespace {

using Bytes = std::vector<std::uint8_t>;
using RoundTrip = std::function<DecodeError(const Bytes&, Bytes&)>;

Writer shared_writer(1);

template <typename Codec, typename Model>
RoundTrip round_trip() {
    return [](const Bytes& payload, Bytes& out) {
        Model value{};
        Reader reader(payload.data(), payload.size());
        DecodeError error = Codec::decode(reader, value);
        if (!error.ok()) return error;
        Writer writer;
        Codec::encode(writer, value);
        out = writer.bytes();
        return error;
    };
}

template <typename Codec, typename Model>
RoundTrip shared_round_trip() {
    return [](const Bytes& payload, Bytes& out) {
        Model value{};
        Reader reader(payload.data(), payload.size());
        DecodeError error = Codec::decode(reader, value);
        if (!error.ok()) return error;
        shared_writer.clear();
        Codec::encode(shared_writer, value);
        out = shared_writer.bytes();
        return error;
    };
}

std::size_t reused_allocations = 0;

template <typename Codec, typename Model>
RoundTrip reused_round_trip() {
    return [](const Bytes& payload, Bytes& out) {
        static Model held{};
        Reader reader(payload.data(), payload.size());
        DecodeError error = Codec::decode(reader, held);
        if (!error.ok()) return error;
        shared_writer.clear();
        Codec::encode(shared_writer, held);
        out = shared_writer.bytes();
        std::size_t before = allocations;
        Reader again(payload.data(), payload.size());
        error = Codec::decode(again, held);
        reused_allocations = allocations - before;
        return error;
    };
}

const std::map<std::string, RoundTrip> kReusedRoundTrips = {
    {"Player.edge", reused_round_trip<PlayerEdgeCodec, models::Player>()},
    {"DeviceState.edge", reused_round_trip<DeviceStateEdgeCodec, models::DeviceState>()},
    {"DeviceState.unity", reused_round_trip<DeviceStateUnityCodec, models::DeviceState>()},
    {"EveryPrimitive.edge", reused_round_trip<EveryPrimitiveEdgeCodec, models::EveryPrimitive>()},
    {"Team.edge", reused_round_trip<TeamEdgeCodec, models::Team>()},
};

const std::map<std::string, RoundTrip> kSharedRoundTrips = {
    {"Player.edge", shared_round_trip<PlayerEdgeCodec, models::Player>()},
    {"DeviceState.edge", shared_round_trip<DeviceStateEdgeCodec, models::DeviceState>()},
    {"DeviceState.unity", shared_round_trip<DeviceStateUnityCodec, models::DeviceState>()},
    {"EveryPrimitive.edge", shared_round_trip<EveryPrimitiveEdgeCodec, models::EveryPrimitive>()},
    {"Team.edge", shared_round_trip<TeamEdgeCodec, models::Team>()},
};

const std::map<std::string, RoundTrip> kRoundTrips = {
    {"Player.edge", round_trip<PlayerEdgeCodec, models::Player>()},
    {"DeviceState.edge", round_trip<DeviceStateEdgeCodec, models::DeviceState>()},
    {"DeviceState.unity", round_trip<DeviceStateUnityCodec, models::DeviceState>()},
    {"EveryPrimitive.edge", round_trip<EveryPrimitiveEdgeCodec, models::EveryPrimitive>()},
    {"Team.edge", round_trip<TeamEdgeCodec, models::Team>()},
};

struct Identity {
    std::uint32_t id;
    std::uint64_t fingerprint;
};

const std::map<std::string, Identity> kIdentities = {
    {"Player.edge", {PlayerEdgeCodec::kMessageId, PlayerEdgeCodec::kFingerprint}},
    {"PlayerInfo.edge", {PlayerInfoEdgeCodec::kMessageId, PlayerInfoEdgeCodec::kFingerprint}},
    {"DeviceState.edge", {DeviceStateEdgeCodec::kMessageId, DeviceStateEdgeCodec::kFingerprint}},
    {"DeviceState.unity", {DeviceStateUnityCodec::kMessageId, DeviceStateUnityCodec::kFingerprint}},
    {"EveryPrimitive.edge", {EveryPrimitiveEdgeCodec::kMessageId, EveryPrimitiveEdgeCodec::kFingerprint}},
    {"Team.edge", {TeamEdgeCodec::kMessageId, TeamEdgeCodec::kFingerprint}},
};

const std::map<std::string, DecodeError::Kind> kErrorKinds = {
    {"UnexpectedEof", DecodeError::Kind::UnexpectedEof},
    {"InvalidBool", DecodeError::Kind::InvalidBool},
    {"InvalidUtf8", DecodeError::Kind::InvalidUtf8},
    {"LengthOverflow", DecodeError::Kind::LengthOverflow},
};

int checks = 0;
int failures = 0;

void check(bool passed, const std::string& failure) {
    ++checks;
    if (!passed) {
        ++failures;
        std::printf("FAIL %s\n", failure.c_str());
    }
}

Bytes from_hex(const std::string& text) {
    Bytes out;
    if (text == "-") return out;
    for (std::size_t i = 0; i + 1 < text.size(); i += 2) {
        out.push_back(static_cast<std::uint8_t>(std::stoul(text.substr(i, 2), nullptr, 16)));
    }
    return out;
}

std::string to_hex(const Bytes& bytes) {
    std::string out;
    char digits[3];
    for (std::uint8_t byte : bytes) {
        std::snprintf(digits, sizeof digits, "%02x", byte);
        out += digits;
    }
    return out;
}

}

void check_net_schema() {
    check(fmx_schema_check(&FOMOXA_NET_SCHEMA) == FMX_OK, "net schema: fmx_schema_check rejected it");
    check(FOMOXA_NET_SCHEMA.fingerprint == FOMOXA_SCHEMA_FINGERPRINT, "net schema: fingerprint differs from the handshake");
    check(FOMOXA_NET_SCHEMA.message_count == FOMOXA_MESSAGES.size(), "net schema: message count differs from the handshake");
    for (const FomoxaMessage& expected : FOMOXA_MESSAGES) {
        const fmx_message_schema* found = fmx_schema_message(&FOMOXA_NET_SCHEMA, expected.id);
        bool matches = found != nullptr && found->fingerprint == expected.fingerprint &&
                       found->prefix_count == expected.prefix_count &&
                       std::equal(expected.prefixes, expected.prefixes + expected.prefix_count, found->prefixes);
        check(matches, std::string("net schema: ") + expected.name + " does not match the handshake table");
    }
}

int main() {
    std::string line;
    while (std::getline(std::cin, line)) {
        std::istringstream fields(line);
        std::string kind;
        std::string name;
        fields >> kind >> name;

        if (kind == "message") {
            std::string id_text;
            std::string fingerprint_text;
            fields >> id_text >> fingerprint_text;
            auto expected = kIdentities.find(name);
            if (expected == kIdentities.end()) continue;
            std::uint32_t id = static_cast<std::uint32_t>(std::stoul(id_text, nullptr, 16));
            std::uint64_t fingerprint = std::stoull(fingerprint_text, nullptr, 16);
            check(id == expected->second.id, name + ": message id mismatch");
            check(fingerprint == expected->second.fingerprint, name + ": fingerprint mismatch");
            continue;
        }

        std::string message;
        std::string payload_text;
        std::string last;
        fields >> message >> payload_text >> last;
        Bytes payload = from_hex(payload_text);
        Bytes actual;
        DecodeError error = kRoundTrips.at(message)(payload, actual);

        if (kind == "accept") {
            Bytes expected = from_hex(last);
            check(error.ok(), "accept " + name + ": " + error.message());
            check(!error.ok() || actual == expected,
                  "accept " + name + ": re-encoded " + to_hex(actual) + ", expected " + to_hex(expected));
            Bytes shared;
            DecodeError shared_error = kSharedRoundTrips.at(message)(payload, shared);
            check(shared_error.ok() && shared == expected,
                  "shared writer " + name + ": " + to_hex(shared) + ", expected " + to_hex(expected));
            Bytes reused;
            DecodeError reused_error = kReusedRoundTrips.at(message)(payload, reused);
            check(reused_error.ok() && reused == expected,
                  "reused target " + name + ": " + to_hex(reused) + ", expected " + to_hex(expected));
            check(reused_allocations == 0,
                  "reused target " + name + ": decoding the same bytes again allocated " + std::to_string(reused_allocations) + " times");
        } else if (kind == "reject") {
            check(!error.ok() && error.kind == kErrorKinds.at(last),
                  "reject " + name + ": " + (error.ok() ? std::string("decoded") : error.message()) + ", expected " + last);
        }
    }

    check_net_schema();

    std::printf("%d/%d checks passed\n", checks - failures, checks);
    return failures == 0 && checks > 0 ? 0 : 1;
}

File: tests/fixtures-ts/vectors_test.ts
import * as fs from "node:fs";
import * as path from "node:path";

import { Player, Team } from "./src/models/player";
import { DeviceState } from "./src/models/device_state";
import { EveryPrimitive } from "./src/models/every_primitive";
import { DecodeError, Reader, Writer } from "./src/generated/runtime";
import { PlayerEdgeCodec } from "./src/generated/player_edge";
import { PlayerInfoEdgeCodec } from "./src/generated/player_info_edge";
import { TeamEdgeCodec } from "./src/generated/team_edge";
import { DeviceStateEdgeCodec } from "./src/generated/device_state_edge";
import { DeviceStateUnityCodec } from "./src/generated/device_state_unity";
import { EveryPrimitiveEdgeCodec } from "./src/generated/every_primitive_edge";
import { FOMOXA_MESSAGES, FOMOXA_SCHEMA_FINGERPRINT } from "./src/generated/handshake";
import { fomoxaNetSchema } from "./src/generated/net_schema";

interface Codec<T> {
    MESSAGE_ID: number;
    FINGERPRINT: bigint;
    encode(writer: Writer, value: T): void;
    decode(reader: Reader, value: T): void;
}

interface Vector {
    name: string;
    message: string;
    hex: string;
    reencode?: string;
    error?: string;
}

const sharedWriter = new Writer(1);

function roundTrip<T>(codec: Codec<T>, create: () => T): (payload: Uint8Array) => Uint8Array {
    return (payload) => {
        const value = create();
        codec.decode(new Reader(payload), value);
        const writer = new Writer();
        codec.encode(writer, value);
        return writer.toUint8Array();
    };
}

function sharedRoundTrip<T>(codec: Codec<T>, create: () => T): (payload: Uint8Array) => Uint8Array {
    return (payload) => {
        const value = create();
        codec.decode(new Reader(payload), value);
        sharedWriter.clear();
        codec.encode(sharedWriter, value);
        return sharedWriter.writtenView();
    };
}

const roundTrips: Record<string, (payload: Uint8Array) => Uint8Array> = {
    "Player.edge": roundTrip(PlayerEdgeCodec, () => new Player()),
    "DeviceState.edge": roundTrip(DeviceStateEdgeCodec, () => new DeviceState()),
    "DeviceState.unity": roundTrip(DeviceStateUnityCodec, () => new DeviceState()),
    "EveryPrimitive.edge": roundTrip(EveryPrimitiveEdgeCodec, () => new EveryPrimitive()),
    "Team.edge": roundTrip(TeamEdgeCodec, () => new Team()),
};

const sharedRoundTrips: Record<string, (payload: Uint8Array) => Uint8Array> = {
    "Player.edge": sharedRoundTrip(PlayerEdgeCodec, () => new Player()),
    "DeviceState.edge": sharedRoundTrip(DeviceStateEdgeCodec, () => new DeviceState()),
    "DeviceState.unity": sharedRoundTrip(DeviceStateUnityCodec, () => new DeviceState()),
    "EveryPrimitive.edge": sharedRoundTrip(EveryPrimitiveEdgeCodec, () => new EveryPrimitive()),
    "Team.edge": sharedRoundTrip(TeamEdgeCodec, () => new Team()),
};

const identities: Record<string, { MESSAGE_ID: number; FINGERPRINT: bigint }> = {
    "Player.edge": PlayerEdgeCodec,
    "PlayerInfo.edge": PlayerInfoEdgeCodec,
    "DeviceState.edge": DeviceStateEdgeCodec,
    "DeviceState.unity": DeviceStateUnityCodec,
    "EveryPrimitive.edge": EveryPrimitiveEdgeCodec,
    "Team.edge": TeamEdgeCodec,
};

const errorPrefixes: Record<string, string> = {
    UnexpectedEof: "unexpected eof",
    InvalidBool: "invalid bool",
    InvalidUtf8: "invalid utf-8",
    LengthOverflow: "length overflow",
};

let checks = 0;
let failures = 0;

function check(passed: boolean, failure: string): void {
    checks += 1;
    if (!passed) {
        failures += 1;
        console.log(`FAIL ${failure}`);
    }
}

function hex(text: string): Uint8Array {
    return Uint8Array.from(Buffer.from(text.replace(/\s+/g, ""), "hex"));
}

function toHex(bytes: Uint8Array): string {
    return Buffer.from(bytes).toString("hex");
}

function describe(error: unknown): string {
    return error instanceof Error ? `${error.name}: ${error.message}` : String(error);
}

const vectorsPath = process.argv[2] ?? path.join(__dirname, "..", "..", "vectors", "fomoxa-vectors.json");
const vectors = JSON.parse(fs.readFileSync(vectorsPath, "utf8"));

for (const [name, codec] of Object.entries(identities)) {
    const message = vectors.messages[name];
    const id = Number.parseInt(message.id.slice(2), 16);
    const fingerprint = BigInt(`0x${message.fingerprint.slice("sha256:".length, "sha256:".length + 16)}`);
    check(codec.MESSAGE_ID === id, `${name}: message id ${codec.MESSAGE_ID.toString(16)}, vectors say ${id.toString(16)}`);
    check(codec.FINGERPRINT === fingerprint, `${name}: fingerprint ${codec.FINGERPRINT.toString(16)}, vectors say ${fingerprint.toString(16)}`);
}

for (const vector of vectors.accept as Vector[]) {
    const expected = hex(vector.reencode ?? "");
    try {
        const actual = roundTrips[vector.message](hex(vector.hex));
        check(toHex(actual) === toHex(expected), `accept ${vector.name}: re-encoded ${toHex(actual)}, expected ${toHex(expected)}`);
    } catch (error) {
        check(false, `accept ${vector.name}: ${describe(error)}`);
    }
}

for (const vector of vectors.reject as Vector[]) {
    const expected = vector.error ?? "";
    try {
        roundTrips[vector.message](hex(vector.hex));
        check(false, `reject ${vector.name}: decoded, expected ${expected}`);
    } catch (error) {
        const matched = error instanceof DecodeError && error.message.startsWith(errorPrefixes[expected]);
        check(matched, `reject ${vector.name}: ${describe(error)}, expected DecodeError ${expected}`);
    }
}

for (const vector of vectors.accept as Vector[]) {
    const expected = toHex(roundTrips[vector.message](hex(vector.hex)));
    const actual = toHex(sharedRoundTrips[vector.message](hex(vector.hex)));
    check(actual === expected, `shared writer ${vector.name}: ${actual}, expected ${expected}`);
}

function fingerprint64(tagged: string): bigint {
    return BigInt(`0x${tagged.slice("sha256:".length, "sha256:".length + 16)}`);
}

try {
    const schema = fomoxaNetSchema();
    check(schema.fingerprint === FOMOXA_SCHEMA_FINGERPRINT, `net schema: fingerprint ${schema.fingerprint.toString(16)}, handshake says ${FOMOXA_SCHEMA_FINGERPRINT.toString(16)}`);
    check(schema.messages.length === FOMOXA_MESSAGES.length, `net schema: ${schema.messages.length} messages, handshake declares ${FOMOXA_MESSAGES.length}`);
    for (const [name, message] of Object.entries(vectors.messages) as [string, { id: string; fingerprint: string; prefixes: string[] }][]) {
        const found = schema.byId.get(Number.parseInt(message.id.slice(2), 16));
        const expected = message.prefixes.map(fingerprint64);
        const matches =
            found !== undefined &&
            found.fingerprint === fingerprint64(message.fingerprint) &&
            found.prefixes.length === expected.length &&
            found.prefixes.every((prefix, index) => prefix === expected[index]);
        check(matches, `net schema: ${name} does not match the vectors`);
    }
} catch (error) {
    check(false, `net schema: ${describe(error)}`);
}

console.log(`${checks - failures}/${checks} checks passed`);
process.exitCode = failures === 0 ? 0 : 1;
